src/utils/i18n.ts

```diff
@@ -0,0 +1,264 @@
+ *                          available.
+ * @returns                 The first locale that exactly matches or has a subset that matches.
+ *                          Results can be undefined if no match is found.
+ */
+export function localeLookup(
+  preferredLocales: string | string[],
+  supportedLocales: string[],
+  aliases?: Aliases
+): string | undefined {
+  // Normalize all the locales.
+  const normalizedPreferredlocales = getCanonicalLocales(preferredLocales);
+  const supportedSet: Set = {};
+
+  getCanonicalLocales(supportedLocales).forEach((l) => (supportedSet[l] = true));
+
+  // Normalize aliases.
+  const normalizedAliases: Aliases = {};
+
+  if (isAlias(aliases)) {
+    for (let key in aliases) {
+      const value = aliases[key];
+
+      if (typeof value === 'string') {
+        const canonicalLocale = getCanonicalLocale(key);
+        normalizedAliases[canonicalLocale] = getCanonicalLocale(value);
+      }
+    }
+  }
+
+  // Since aliases may or may not be normalized,
+  // If there is no match found while iterating through the list, use this one.
+  const possibleLocales: string[] = [];
+
+  // Create a new list of locales that will increase the odds of matching.
+  const baseSet: Set = {};
+
+  // Add locales to the base set.
+  normalizedPreferredlocales.forEach((l) => (baseSet[l] = true));
+
+  // Create list of possible locales.
+  for (let i = 0; i < normalizedPreferredlocales.length; i++) {
+    const current = normalizedPreferredlocales[i];
+
+    possibleLocales.push(current);
+
+    // Create subsets of current and add them to the list of possible locales.
+    // Only add them if they aren't subsets of the next locale or if they appear later in the
+    // preference list.
+    const next = normalizedPreferredlocales[i + 1] || '';
+    const subsets = makeSubsets(current).filter((s) => !baseSet[s] && next.search(s) === -1);
+
+    possibleLocales.push(...subsets);
+  }
+
+  const matchedLocales =
+    possibleLocales
+      // Swap any locales that have aliases with their alias.
+      .map((l) => {
+        const alias = normalizedAliases[l];
+        return typeof alias === 'string' ? alias : l;
+      })
+      // Filter for a locale that is in the supported set.
+      .filter((l) => supportedSet[l] === true);
+
+  // Return the first locale that matched (or undefined if it is not available).
+  return matchedLocales[0];
+}
+
+function isAlias(aliases: Aliases | undefined): aliases is Aliases {
+  return typeof aliases === 'object';
+}
+
+function isTranslations(translations: Translations[string]): translations is Translations {
+  return typeof translations === 'object';
+}
+
+/**
+ * Generates viable subsets from a given locale.
+ *
+ * Examples:
+ * ```
+ * new LanguageTag('zh-Hant-HK').makeSubsets()    // ['zh-Hant', 'zh']
+ * new LanguageTag('zh-HK').makeSubsets()         // ['zh']
+ * new LanguageTag('zh-Hant').makeSubsets()       // ['zh']
+ * new LanguageTag('zh').makeSubsets()            // []
```

Comment:
Are these considered subsets of the locale? I don't know the terminology that the RFC uses, but they seem more like supersets at first glance, since `zh` is more generic/broad than `zh-Hant-HK`, for example.

Reply:
You are right, actually. I never considered it that way. I just thought of it more as being the subset of the strings.

I suppose I could call it `makeLanguageRangeSet`.